Subdomain basic statistics .

aalfonsi opened the pull request:
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #1923 

##### What are the significant changes in functionality due to this change request?
Added a new post processor to compute subdomain basic statistics leveraging RAVEN basic statistics. 

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

aalfonsi commented on ravenframework/Models/PostProcessors/BasicStatistics.py:
```diff
@@ -1551,7 +1562,8 @@ def run(self, inputIn):
     inputData = self.inputToInternal(inputIn)
-    outputSet = self.__runLocal(inputData)
+    outputSet = self._runLocal(inputData)
+    print(outputSet)
```
```suggestion
```

mandd commented:
I have left my initial set of comments. One of my main questions is why a new PP class was added rather than extending the directly the basic statistics?

aalfonsi commented on doc/user_manual/PostProcessors/SubdomainBasicStatistics.tex:
```diff
@@ -0,0 +1,112 @@
+
+In this case, the RAVEN variables ``mean\_x01, mean\_x02, sen\_x01\_a, sen\_x02\_a, sen\_x01\_b, sen\_x02\_b''
+will be computed for each cell defined above (2 cells in this case).
+The results will be associated to the mid-point of each point. For the example above:
```
```suggestion
The results will be associated to the mid-point of each cell. For the example above:
```

aalfonsi commented on doc/user_manual/PostProcessors/SubdomainBasicStatistics.tex:
```diff
@@ -0,0 +1,112 @@
+In this case, the RAVEN variables ``mean\_x01, mean\_x02, sen\_x01\_a, sen\_x02\_a, sen\_x01\_b, sen\_x02\_b''
+will be computed for each cell defined above (2 cells in this case).
+The results will be associated to the mid-point of each point. For the example above:
+\begin{itemize}
```
The interval of the fist grid is
[10,100] with 2 steps:
`|---------|---------|`
`10        55       100`
midpoints:
 `------|--------|------`
 `      32.5     77.5 `

mandd replied:
gotcha

aalfonsi commented on doc/user_manual/raven_user_manual.tex:
```diff
@@ -90,6 +90,19 @@
+\setlist[itemize,7]{label=$\bullet$}
+\setlist[itemize,8]{label=$\bullet$}
+\renewlist{itemize}{itemize}{8}
+
```
I added a comment at the begin of the command:
```the following allows for 8 levels list nesting```

This allows to add an additional level of nesting for the "itemize" command

aalfonsi commented on doc/user_manual/sampler.tex:
```diff
@@ -118,6 +118,34 @@ \section{Samplers}
+    For example, if the grid \xmlAttr{type} is \xmlString{value}, in the body
+    of \xmlNode{grid}, the user will specify the values representing the nodes of the grid.
+\end{itemize}
+}
```
yeah probably not the right location. Moved to postprocessors.tex

aalfonsi commented on doc/user_manual/sampler.tex:
```diff
@@ -234,6 +262,21 @@ \section{Samplers}
 }
 
+\newcommand{\gridDescriptionNoCFD}
+{
```
Typo. CFD -> CDF (Cumulative Dist Function)

aalfonsi commented on doc/user_manual/PostProcessors/SubdomainBasicStatistics.tex:
```diff
@@ -0,0 +1,112 @@
+
+In this case, the RAVEN variables ``mean\_x01, mean\_x02, sen\_x01\_a, sen\_x02\_a, sen\_x01\_b, sen\_x02\_b''
+will be computed for each cell defined above (2 cells in this case).
+The results will be associated to the mid-point of each point. For the example above:
```
fixed

aalfonsi commented:
> I have left my initial set of comments. One of my main questions is why a new PP class was added rather than extending the directly the basic statistics?

@mandd addressed your comments. The reason is mostly because it would have required an additional refactoring of the basic statistics and it is already quite complex. The advantage of having this PP is that the computation is encapsulated in the BasicStatistics and, in the future, the calculation for each "cell" can be parallelized if needed.

wangcj05 commented:
Hi @aalfonsi: Is it possible that the subdomain information (the grid lower and upper bound) also get stored in the DataObjects, for example in the metadata? I would recommend to preserve this information in the output DataObject.

aalfonsi commented:
> Hi @aalfonsi: Is it possible that the subdomain information (the grid lower and upper bound) also get stored in the DataObjects, for example in the metadata? I would recommend to preserve this information in the output DataObject.

@wangcj05 can you open an issue and I can address it in a followup MR?
